### ISSUE TITLE
Performing a MySQL restore breaks wal-g binlog-push

### ISSUE THREAD

**Issue by jstaf** [wal-g/wal-g#889]
If you have a server that is backed up by WAL-G (and are using `wal-g binlog-push` to archive binlogs), restoring a backup using WAL-G will permanently break future binlog archival for that server.

To reproduce:

* Create a new MySQL server and import some data on it.
* Run `FLUSH BINARY LOGS;` several times to rotate the binlog. (If binlogs are named using the format `binlog.######`, you should have higher-numbered binlogs like `binlog.000005` now if you run `SHOW BINARY LOGS;`
* Archive binlogs with `wal-g binlog-push` (all binlogs except for current one get archived)
* Restore database from backups
* Newly restored database starts over at `binlog.000001` (check via `SHOW BINARY LOGS;`).
* Run `FLUSH BINARY LOGS;` once so `wal-g binlog-push` can archive the new `binlog.000001` binlog.
* Running `wal-g binlog-push` no longer works, even though there's new data in `binlog.000001`. This binlog will never be archived by `wal-g binlog-push`, breaking point-in-time restores going forwards.

It looks like WAL-G is incorrectly refusing to sync binlogs if there is a higher-numbered binlog in cloud storage. I've done some testing and this happens irrespective of if there's an existing `binlog.000001` file or not in the cloud storage. As long as there's a higher numbered binlog in cloud storage, `wal-g binlog-push` won't do anything.

The correct behavior here would be to always sync binlogs if there's not a binlog with the same name already in cloud storage, and the binlog on the local MySQL server is newer than the modification time on S3 or wherever.

---------

As a workaround until this is fixed you can run something like this to trick the MySQL server into starting new binlogs at the "highest numbered binlog+1" post restore (so `wal-g binlog-push` will keep working following a restore):

```bash
# after restoring a backup via wal-g backup-fetch, but BEFORE starting the mysql server:

# change .br to .lz4 if using lz4 instead of brotli
touch "/var/lib/mysql/$(aws s3 ls "$WALG_S3_PREFIX/binlog_005/" | tail -n 1 | awk '{ print $4}' | sed 's/.br//g')"
```

**Comment by x4m** [comment]
Thanks for pointing this out.
I think the problem arose due to the absence of timelines. We need to distinguish binlogs from different "incarnations" of the same node in the cluster.
@mialinx can you please take a look? I believe this can affect our resetup machinery too.

**Comment by mialinx** [comment]
Wal-g expects (at the moment) that every server have different binlog naming patter, in order to continue binlog archiving after failover.
After restoring  from backup you get a new server actually.

So you just set different binlog name in my.cnf for a server after restoring:
```
log_bin = binlog-seq-1
log_bin = binlog-seq-2
log_bin = binlog-seq-3
etc...
```

The correct behaviour is a bit more complex... 
We can't just replace already uploaded binlog (as it breaks PITR).
We can't (as you shown) rely on binlog naming in MySQL instances.
So we need to upload binary logs with different names and track check the latest uploaded..

**Comment by jstaf** [comment]
What about using the backup UUID from `xtrabackup_info` as a timeline identifier? When performing a restore, xtrabackup will drop a `xtrabackup_info` file with a UUID for the backup into the mysql datadir. This would work as a unique identifier for the MySQL backup timeline (if `xtrabackup_info` is present, you can use that file to get the timeline name for binlogs, and if the file isn't present, it means we can store binlogs under the default name format). 

The only downside to this approach is that it wouldn't work for workflows using `mysqldump` because they don't create `xtrabackup_info` files. An alternative approach could just be storing the name a backup was restored from (for instance: stream_123456T789Z) in the MySQL database itself post-restore (xtrabackup does something similar for tracking backup history for incremental backups: https://www.percona.com/doc/percona-xtrabackup/LATEST/innobackupex/storing_history.html), or in a file somewhere similar to `xtrabackup_info`.
